blockchain/blockindex.go

```diff
@@ -574,14 +869,417 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// recent tips when possible.
+	bi.forEachChainTipAfterHeight(node, func(tip *blockNode) error {
+		markDescendantsFailed(node, tip)
+		return nil
+	})
+
+	// Update the best header if the current one is now invalid which will be
+	// the case when the best header is a descendant of the failed block.
+	if bi.bestHeader.status.KnownInvalid() {
+		// Use the first ancestor of the failed block that is not known to be
+		// invalid as the lower bound for the best header.  This will typically
+		// be parent of the failed block, but it might be some more distant
+		// ancestor when performing manual invalidation.
+		n := node.parent
+		for n != nil && n.status.KnownInvalid() {
+			n = n.parent
+		}
+		bi.bestHeader = n
+
+		// Scour the block tree to find a new best header.
+		//
+		// Note that all chain tips must be iterated versus filtering based on
+		// the current best header height because, while uncommon, it is
+		// possible for lower heights to have more work.
+		bi.forEachChainTip(func(tip *blockNode) error {
+			// Skip chain tips that are descendants of the failed block since
+			// none of the intermediate headers are eligible to become the best
+			// header given they all have an invalid ancestor.
+			if tip.Ancestor(node.height) == node {
+				return nil
+			}
+
+			bi.maybeUpdateBestHeaderForTip(tip)
+			return nil
+		})
+	}
+	bi.Unlock()
+}
+
+// canValidate returns whether or not the block associated with the provided
+// node can be validated.  In order for a block to be validated, both it, and
+// all of its ancestors, must have the block data available.
+//
+// This function MUST be called with the block index lock held (for reads).
+func (bi *blockIndex) canValidate(node *blockNode) bool {
+	return node.isFullyLinked && node.status.HaveData()
+}
+
+// CanValidate returns whether or not the block associated with the provided
+// node can be validated.  In order for a block to be validated, both it, and
+// all of its ancestors, must have the block data available.
+//
+// This function is safe for concurrent access.
+func (bi *blockIndex) CanValidate(node *blockNode) bool {
+	bi.RLock()
+	canValidate := bi.canValidate(node)
+	bi.RUnlock()
+	return canValidate
+}
+
+// removeLessWorkCandidates removes all potential best chain candidates that
+// have less work than the provided node, which is typically a newly connected
+// best chain tip.
+//
+// This function MUST be called with the block index lock held (for writes).
+func (bi *blockIndex) removeLessWorkCandidates(node *blockNode) {
+	// Remove all best chain candidates that have less work than the passed
+	// node.
+	for n := range bi.bestChainCandidates {
+		if n.workSum.Cmp(node.workSum) < 0 {
+			bi.removeBestChainCandidate(n)
+		}
+	}
+
+	// The best chain candidates must always contain at least the current best
+	// chain tip.  Assert this assumption is true.
+	if len(bi.bestChainCandidates) == 0 {
```

Comment:
Perhaps an exception can be made in the work sum comparisons to not remove the current best chain tip if it is ha less work than the passed node to mitigate panicking since the list will be guaranteed to always have one entry?

Reply:
The best chain tip is only know to the chain instance itself as opposed to the block index, so it would have to be passed in.  At that point, you'd end up with the same potential issue of it being incorrect and would ideally want to assert that too, so I don't think it would buy anything.